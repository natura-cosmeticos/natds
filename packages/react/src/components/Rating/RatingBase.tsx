import React from 'react'
import { Size, Theme } from '@naturacosmeticos/natds-themes'
import { createUseStyles } from 'react-jss'
import Icon from '../Icon'
import IconButtonBase from '../IconButton/IconButtonBase'

export interface RatingBaseProps {
  ariaLabel: string
  disabled?: boolean
  iconActive?: boolean
  iconFilled?: boolean
  isClickable?: boolean
  onClick: (e: any) => void
  onMouseEnter?: () => void
  onMouseLeave?: () => void
  size: keyof Size
  testID?: string
  value?: number
}

type RatingBaseStyleProps = Pick<RatingBaseProps, 'disabled' | 'isClickable' | 'iconFilled' | 'iconActive'>

export const getRatingColor = ({ color }: Theme) => (
  { iconFilled, iconActive, disabled }: RatingBaseStyleProps
) => (iconFilled && iconActive && !disabled ? '#F8B546' : color.mediumEmphasis)

const styles = createUseStyles((theme: Theme) => ({
  ratingContainer: {
    display: 'flex'
  },
  ratingIcon: {
    backgroundColor: 'transparent',
    position: 'relative',
    '&:after': {
      backgroundColor: theme.color.highlight,
      borderRadius: '50%',
      content: '" "',
      height: '100%',
      left: 0,
      opacity: 0,
      position: 'absolute',
      top: 0,
      width: '100%'
    },
    '&:hover:not([disabled]):after': { opacity: theme.opacity.lower },
    '&:focus:not([disabled]):after': { opacity: theme.opacity.low },
    '& > i': {
      cursor: ({ disabled, isClickable }: RatingBaseStyleProps) => isClickable && !disabled && 'pointer',
      color: (getRatingColor(theme))
    }
  }
}))

const RatingBase = React.forwardRef<HTMLButtonElement, RatingBaseProps>(({
  ariaLabel,
  disabled = false,
  iconActive = true,
  iconFilled = true,
  isClickable = false,
  onClick,
  onMouseEnter,
  onMouseLeave,
  value,
  size,
  testID
}, ref) => {
  const { ratingIcon, ratingContainer } = styles({
    disabled,
    isClickable,
    iconFilled,
    iconActive
  })

  const checkIcon = iconFilled ? 'filled-action-rating' : 'outlined-action-rating'

  return (
    <div onMouseEnter={onMouseEnter} onMouseLeave={onMouseLeave} className={ratingContainer}>
      <IconButtonBase
        ref={ref}
        ariaLabel={ariaLabel}
        disabled={disabled || !isClickable}
        IconComponent={(
          <Icon
            ariaHidden
            ariaLabel={ariaLabel}
            name={checkIcon}
            role="button"
            size={size}
          />
        )}
        onClick={onClick}
        size={size}
        className={ratingIcon}
        value={value}
        testID={testID}
      />
    </div>
  )
})

export default RatingBase
