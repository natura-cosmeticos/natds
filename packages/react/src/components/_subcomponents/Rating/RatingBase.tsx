import React from 'react'
import { Size, Theme } from '@naturacosmeticos/natds-themes'
import { createUseStyles } from 'react-jss'
import Icon from '../../Icon'
import IconButtonBase from '../IconButton'

export interface RatingBaseProps {
  ariaLabel: string
  disabled?: boolean
  iconActive?: boolean
  iconFilled?: boolean
  isClickable?: boolean
  onClick: () => void
  size: keyof Size
}

export const getRatingColor = ({ color }: Theme) => (
  { iconFilled, iconActive, disabled }: RatingBaseProps
) => (iconFilled && iconActive && !disabled ? '#F8B546' : color.mediumEmphasis)

const styles = createUseStyles((theme: Theme) => ({
  rating: {
    backgroundColor: 'transparent',
    '& > i': {
      display: 'flex',
      color: getRatingColor(theme)
    }
  }
}))

const RatingBase = ({
  ariaLabel,
  disabled = false,
  iconActive,
  iconFilled,
  isClickable,
  onClick,
  size
}: RatingBaseProps): JSX.Element => {
  const { rating } = styles({ iconFilled, iconActive, disabled })
  const checkIcon = iconFilled ? 'filled-action-rating' : 'outlined-action-rating'

  return (
    <IconButtonBase
      classes={rating}
      disabled={disabled}
      size={size}
      onClick={onClick}
      IconComponent={(
        <Icon
          size={size}
          name={checkIcon}
          ariaLabel={ariaLabel}
        />
      )}
    />
  )
}

export default RatingBase
