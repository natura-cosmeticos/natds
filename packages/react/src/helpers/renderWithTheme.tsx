import React from 'react'
import { JssProvider, SheetsRegistry } from 'react-jss'
import { render } from '@testing-library/react'
import { ThemeProvider } from '../ThemeProvider'

const renderWithTheme = (component) => {
  const sheets = new SheetsRegistry()

  const renderTree = render(
    <JssProvider registry={sheets}>
      <ThemeProvider>
        {component}
      </ThemeProvider>
    </JssProvider>
  )

  return { styles: sheets.toString(), component: renderTree }
}

export default renderWithTheme
