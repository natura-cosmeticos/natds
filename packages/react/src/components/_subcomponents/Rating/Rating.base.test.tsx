import React from 'react'
import renderWithTheme from '../../../helpers/renderWithTheme'
import RatingBase from './Rating.base'

describe('RatingBase component', () => {
  it('should render correctly with the given props', () => {
    const { styles, component } = renderWithTheme(<RatingBase ariaLabel="Rating Base" size="large" onClick={() => ''} />)

    expect([styles.toString(), component.container]).toMatchSnapshot()
  })
})
