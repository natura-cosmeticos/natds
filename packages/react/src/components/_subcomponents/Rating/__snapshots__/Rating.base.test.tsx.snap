// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatingBase component should render correctly with the given props 1`] = `
Array [
  ".icon-0-2-15 {
  font-family: natds-icons;
  user-select: none;
}
.icon-d0-0-2-16 {
  color: #000000;
  font-size: 64px;
}
.sharedRippleEffect-0-2-5 {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  content: \\" \\";
  z-index: -1;
  position: absolute;
  border-radius: 50%;
}
.wrapper-0-2-6 {
  outline: none;
  position: relative;
  align-self: start;
}
.wrapper-0-2-6:focus:after {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  content: \\" \\";
  opacity: 0.16;
  z-index: -1;
  position: absolute;
  border-radius: 50%;
}
.wrapper-0-2-6:hover:after {
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  content: \\" \\";
  z-index: -1;
  position: absolute;
  border-radius: 50%;
}
.ripple-0-2-7 {
  opacity: 0;
  position: absolute;
  transform: translate(-50%, -50%);
  border-radius: 50%;
}
.rippleContainer-0-2-8 {
  inset: 0;
  position: absolute;
}
.rippleActive-0-2-9 {
  z-index: 99;
}
@keyframes keyframes-ripple-0-2-10 {
  0% {
    transform: translate(-50%, -50%) scale(0);
  }
  75% {
    opacity: 0.12;
    transform: translate(-50%, -50%) scale(1);
  }
  100% {
    opacity: 0;
    z-index: 0;
  }
}
  .wrapper-d0-0-2-11 {
    cursor: default;
    display: inline-block;
  }
  .wrapper-d0-0-2-11:focus:after {
    background-color: #000000;
  }
  .wrapper-d0-0-2-11:hover:after {
    background-color: #000000;
    opacity: 0;
  }
  .ripple-d1-0-2-12 {
    background-color: #000000;
    height: 0;
    left: 50%;
    top: 50%;
    transition: scale opacity 300ms;
    width: 0;
  }
  .rippleContainer-d2-0-2-13 {
    overflow: hidden;
  }
  .rippleActive-d3-0-2-14 {
    animation: keyframes-ripple-0-2-10 300ms;
  }
  .base-0-2-3 {
    border: none;
    cursor: pointer;
    display: flex;
    outline: none;
    align-items: center;
    justify-content: center;
  }
  .base-0-2-3:disabled {
    cursor: default;
  }
  .base-d0-0-2-4 {
    height: 64px;
    width: 64px;
  }
  .rating-0-2-1 {
    background-color: transparent;
  }
  .rating-0-2-1 > i {
    display: flex;
  }
  .rating-d0-0-2-2 {  }
  .rating-d0-0-2-2 > i {
    color: #777777;
  }",
  <div>
    <div
      class="wrapper-0-2-6 wrapper-d0-0-2-11"
      data-testid="ripple-wrapper"
      tabindex="-1"
    >
      <div
        class="rippleContainer-0-2-8 rippleContainer-d2-0-2-13"
      >
        <div
          class="ripple-0-2-7 ripple-d1-0-2-12 "
          data-testid="ripple-animation"
        />
      </div>
      <button
        class="base-0-2-3 base-d0-0-2-4 rating-0-2-1 rating-d0-0-2-2"
        disabled=""
        type="button"
      >
        <i
          aria-hidden="true"
          arial-label="Rating Base"
          class="icon-0-2-15 icon-d0-0-2-16 natds-icons natds-icons-outlined-action-rating"
          data-testid="icon-outlined-action-rating"
          role="img"
        />
      </button>
    </div>
  </div>,
]
`;
